### PR TITLE
Added batch upsert to WorkspaceService and corresponding API endpoint

### PR DESCRIPTION
**Pull request by MatthewBemis**
Before merging, I'd like to test POSTing some json with upsert instructions in swagger, but there are Google authentication problems preventing me from doing that at the moment. Other than that, this is ready for comments/complaints.

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceApiServiceSpec.scala [hunk 1]
```diff
@@ -243,6 +243,53 @@ class WorkspaceApiServiceSpec extends FlatSpec with HttpService with ScalatestRo
...
+  it should "return 400 when batch upserting an entity with invalid update operations" in withTestDataApiServices { services =>
+    Post(s"/workspaces/${testData.workspace.namespace}/${testData.workspace.name}/entities/batchUpsert", HttpEntity(ContentTypes.`application/json`, Seq(EntityUpdateDefinition("s1", "samples", Seq(RemoveListMember("bingo", AttributeString("a")))), EntityUpdateDefinition("s2", "samples", Seq(AddUpdateAttribute("newAttribute", AttributeString("foo")))), EntityUpdateDefinition("s3", "samples", Seq(RemoveListMember("bingo", AttributeString("a"))))).toJson.toString())) ~>
+      addMockOpenAmCookie ~>
+      sealRoute(services.batchUpsertEntitiesRoute) ~>
+      check {
+        assertResult(StatusCodes.BadRequest) {
+          status
+        }
+        assertResult(2) {
+          println(responseAs[Array[String]].mkString("\n"))
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceApiServiceSpec.scala]: remove the println

File: src/main/scala/org/broadinstitute/dsde/rawls/model/AttributeUpdateOperations.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    }
+  }
+
+  @ApiModel(value = "Entity update definition")
+  case class EntityUpdateDefinition(
+                        @(ApiModelProperty@field)(required = true, value = "The name of the entity")
+                        name: String,
+                        @(ApiModelProperty@field)(required = true, value = "The type of the entity")
+                        entityType: String,
+                        @(ApiModelProperty@field)(required = true, value = "The update operations of the entity")
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/AttributeUpdateOperations.scala]: I'd prefer something like "... to apply to the entity", assuming that is accurate

Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/AttributeUpdateOperations.scala]: That's accurate. I'm actually completely re-working how Swagger is used so I won't fix that in this PR but I'll make sure to update that in the Swagger update PR.

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiService.scala [hunk 3]
```diff
@@ -115,6 +116,35 @@ trait EntityApiService extends HttpService with PerRequestCreator with OpenAmDir
     }
   }
 
+
+  @Path("/batchUpsert")
+  @ApiOperation(value = "batch upsert entities into a workspace",
+    nickname = "batch upsert entities",
+    httpMethod = "Post",
+    produces = "application/json",
+    response = classOf[Entity])
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiService.scala]: Actually returns nothing

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiService.scala [hunk 4]
```diff
@@ -115,6 +116,35 @@ trait EntityApiService extends HttpService with PerRequestCreator with OpenAmDir
...
+  @ApiImplicitParams(Array(
+    new ApiImplicitParam(name = "workspaceNamespace", required = true, dataType = "string", paramType = "path", value = "Workspace Namespace"),
+    new ApiImplicitParam(name = "workspaceName", required = true, dataType = "string", paramType = "path", value = "Workspace Name"),
+    new ApiImplicitParam(name = "batchUpsertJson", required = true, dataType = "org.broadinstitute.dsde.rawls.workspace.AttributeUpdateOperations$EntityUpdateDefinition", paramType = "body", value = "Update operations")
+  ))
+  @ApiResponses(Array(
+    new ApiResponse(code = 204, message = "Successful Request"),
+    new ApiResponse(code = 400, message = "Attribute does not exists or is of an unexpected type"),
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiService.scala]: Is this text correct?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -261,6 +263,33 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, workspaceDAO:
...
+              case Some(e) => e
+              case None => entityDAO.save(workspaceNamespace, workspaceName, Entity(entityUpdate.name, entityUpdate.entityType, Map.empty, workspace.toWorkspaceName), txn)
+            }
+            val trial = Try {
+              val updatedEntity = applyOperationsToEntity(entity, entityUpdate.operations)
+              entityDAO.save(workspaceNamespace, workspaceName, updatedEntity, txn)
+            }
+            (entityUpdate, trial)
+          }
+          val errorMessages = results.collect{
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: This is cool.

File: src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceApiServiceSpec.scala [hunk 6]
```diff
@@ -243,6 +243,52 @@ class WorkspaceApiServiceSpec extends FlatSpec with HttpService with ScalatestRo
       }
   }
 
+  it should "return 400 when batch upserting an entity with invalid update operations" in withTestDataApiServices { services =>
+    Post(s"/workspaces/${testData.workspace.namespace}/${testData.workspace.name}/entities/batchUpsert", HttpEntity(ContentTypes.`application/json`, Seq(EntityUpdateDefinition("s1", "samples", Seq(RemoveListMember("bingo", AttributeString("a")))), EntityUpdateDefinition("s2", "samples", Seq(AddUpdateAttribute("newAttribute", AttributeString("foo")))), EntityUpdateDefinition("s3", "samples", Seq(RemoveListMember("bingo", AttributeString("a"))))).toJson.toString())) ~>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceApiServiceSpec.scala]: Could you move these updates to a  `val` with one update per line, for readability?

**Comment by jmthibault79** [comment]
Review done.  A few comments.

**Comment by MatthewBemis** [comment]
Joel's comments have been addressed. I also removed all of the stuff related to swagger API generation because that's going to be removed soon anyway.

**Comment by jmthibault79** [comment]
Looks good!  Please rebase/squash.

**Pull request merged**